### PR TITLE
refactor(core/jquery-enhanced): migrate $.allTextNodes to vanilla js

### PR DESCRIPTION
**Pull request by sidvishnoi**
As a part of https://github.com/w3c/respec/issues/1699

### CODE REVIEW

File: src/core/utils.js [hunk 1]
```diff
@@ -582,3 +582,27 @@ export function addId(elem, pfx = "", txt = "", noLC = false) {
...
 }
+
+/**
+ * Returns all the descendant text nodes of an element.
+ * @param {Node} elem
+ * @param {Array:String} exclusions node types (localName) to exclude
+ * @returns {Array:String}
+ */
+export function allTextNodes(elem, exclusions) {
+  const excl = new Set(exclusions);
```
Comment (marcoscaceres) [inline src/core/utils.js]: should move this outside, me thinks... no point in creating it over and over again.

File: src/core/utils.js [hunk 2]
```diff
@@ -582,3 +582,27 @@ export function addId(elem, pfx = "", txt = "", noLC = false) {
...
+  const textNodes = [];
+  getTextNodes(elem);
+  return textNodes;
+
+  function getTextNodes(node) {
```
Comment (marcoscaceres) [inline src/core/utils.js]: we probably don't need this function. Can just just inline all this.

File: src/core/utils.js (hunk 1 above)
Comment (sidvishnoi) [reply src/core/utils.js]: make user call with set instead of array?

File: src/core/utils.js [hunk 3]
```diff
@@ -582,3 +582,27 @@ export function addId(elem, pfx = "", txt = "", noLC = false) {
...
+    if (node.nodeType === 1 && excl.has(node.localName.toLowerCase())) return;
```
Comment (marcoscaceres) [inline src/core/utils.js]: while we are here, we should replace the numbers with `Node.WHATEVER`, so we actually know what each number is.

Comment (marcoscaceres) [reply src/core/utils.js]: isn't `node.localName` always in lower case?

File: src/core/utils.js (hunk 2 above)
Comment (marcoscaceres) [reply src/core/utils.js]: oh, sorry, I didn't see that it was recursive! Never mind.

Comment (sidvishnoi) [reply src/core/utils.js]: can make it iterative though

File: src/core/utils.js (hunk 3 above)
Comment (sidvishnoi) [reply src/core/utils.js]: right!

File: src/core/utils.js (hunk 2 above)
Comment (marcoscaceres) [reply src/core/utils.js]: This all still seems silly tho... would this work? 

```
function getTextNodes(el, exclusions) {
  const textNodes = [];
  const walker = document.createTreeWalker(
    el,
    NodeFilter.SHOW_TEXT,
    null,
    false
  );
  let node;
  while ((node = walker.nextNode())) {
    const isExcluded =
      exclusions.includes(node.parentElement.localName) === false;
    textNodes.push(node);
  }
  return textNodes;
}
```

Comment (marcoscaceres) [reply src/core/utils.js]: fixed typo above.

Comment (sidvishnoi) [reply src/core/utils.js]: event better!
*@sidvishnoi always forgets about TreeWalker API*

Comment (marcoscaceres) [reply src/core/utils.js]: How about: 

```JS
function getTextNodes(el, exclusions = []) {
  const acceptNode = node => {
    return exclusions.includes(node.parentElement.localName)
      ? NodeFilter.FILTER_REJECT
      : NodeFilter.FILTER_ACCEPT;
  };
  const nodeIterator = document.createNodeIterator(
    el,
    NodeFilter.SHOW_TEXT,
    { acceptNode },
    false
  );
  const textNodes = [];
  let node;
  while ((node = nodeIterator.nextNode())) {
    textNodes.push(node);
  }
  return textNodes;
}
```

Comment (sidvishnoi) [reply src/core/utils.js]: _@sidvishnoi goes to read the docs_

Comment (marcoscaceres) [reply src/core/utils.js]: Maybe the above... completely untested 🤪

Comment (sidvishnoi) [reply src/core/utils.js]: it does work.
any reason to use NodeIterator instead of TreeWalker?
``` js
  // with TreeWalker
  while (walker.nextNode()) {
    textNodes.push(walker.currentNode);
  }
```

Comment (marcoscaceres) [reply src/core/utils.js]: > NodeIterators and TreeWalkers are two different ways of representing the nodes of a document subtree and a position within the nodes they present. A NodeIterator presents a flattened view of the subtree as an ordered sequence of nodes, presented in document order. Because this view is presented without respect to hierarchy, iterators have methods to move forward and backward, but not to move up and down. Conversely, a TreeWalker maintains the hierarchical relationships of the subtree, allowing navigation of this hierarchy. In general, TreeWalkers are better for tasks in which the structure of the document around selected nodes will be manipulated, while NodeIterators are better for tasks that focus on the content of each selected node.

So, tl;dr: `NodeIterator` is flat, which is what we want.... don't worry, TIL me too!

Comment (marcoscaceres) [reply src/core/utils.js]: > someone should add this example to MDN docs! 

Thanks for volunteering! 🤣

**Comment by marcoscaceres** [comment]
Should probably land each of these one at a time. WDTY?

**Comment by sidvishnoi** [comment]
> Tests

A test for this method exists:
https://github.com/w3c/respec/blob/develop/tests/spec/core/jquery-enhanced-spec.js#L98

core/inlines uses this method and it works.

> Should probably land each of these one at a time. WDTY?

Sure. I'll add the test to utils-specs. And migrate all usage of $.allTextNodes in ReSpec.

Will add deprecation notice to jquery-enhanced when all are done. WDYT?

**Comment by marcoscaceres** [comment]
> Will add deprecation notice to jquery-enhanced when all are done. WDYT?

Sounds good.

**Comment by sidvishnoi** [comment]
ReSpec v21.1.0 now?
